### PR TITLE
Refactor `FusionDefinition` to support deserialization.

### PR DESCRIPTION
**Pull request by rdspring1**
The Fusion Definition class performs several functions:
1. Track `RecordFunctor` objects for math operations
2. Traverses the Trie Structure of Fusion Cache
3. After building a definition, build a CPP Fusion if it does not exist in the Fusion Cache.
4. For user-scheduled fusions, it creates a separate CPP Fusion definition. This definition is modified by the python scheduling operations.

This PR creates the `FusionState` class, which tracks `RecordFunctor` objects and builds a CPP fusion. It doesn't traverse or modify the Fusion Cache. The `FusionDefinition` class now inherits from the `FusionState` class, since the scheduling operations need direct access to the CPP `TensorViews`.

### CODE REVIEW

**Comment by rdspring1** [comment]
The `FusionContainer` will be used in deserialization here. 
https://github.com/csarofeen/pytorch/pull/2478/files#diff-84e340e16680c0d13395ecfbae6b8932ee35548d732c3bbab05cca0d67f805caR383-R385

File: third_party/nvfuser/csrc/python_frontend/fusion_definition.h [hunk 1]
```diff
@@ -72,6 +74,69 @@ struct TORCH_CUDA_CU_API Scalar {
   size_t index;
 };
 
+//! Fusion State contains the information used to build a new cpp Fusion object.
+class FusionState {
+ public:
```
Comment (kevinstephano) [inline third_party/nvfuser/csrc/python_frontend/fusion_definition.h]: I would put the `FusionState` class in a separate file (ie `fusion_state.h` and `fusion_state.cpp`).

File: third_party/nvfuser/csrc/python_frontend/fusion_definition.h [hunk 2]
```diff
@@ -3,18 +3,20 @@
...
 struct UserSchedule;
 
 //! This is helper function used to print a python formated
 //! Fusion IR DataType when printing a fusion definition.
 
-TORCH_CUDA_CU_API const char* dtypeToPyString(nvfuser::PrimDataType t);
+TORCH_CUDA_CU_API const char* dtypeToPyString(PrimDataType t);
 
 //! The State and the StateType enum are used to define state objects to
 //! encapsulate the recording of state in the FusionDefinition.
```
Comment (kevinstephano) [inline third_party/nvfuser/csrc/python_frontend/fusion_definition.h]: I would also put the `StateType` enum and the `State` struct with in your `fusion_state.h` file given I don't believe they are needed by the `FusionDefinition` anymore, correct?

**Review by kevinstephano — APPROVED** [review]
LGTM!

File: third_party/nvfuser/csrc/python_frontend/fusion_definition.cpp [hunk 3]
```diff
@@ -100,42 +88,37 @@ void FusionDefinition::finalizeDefinition() {
...
+    if (isDebugDumpEnabled(DebugDumpOption::FusionIrPresched)) {
       printIr();
     }
   } else {
-    if (nvfuser::isDebugDumpEnabled(
-            nvfuser::DebugDumpOption::PythonFrontendDebug)) {
+    if (isDebugDumpEnabled(DebugDumpOption::PythonFrontendDebug)) {
       std::cout << "\nFusionDefinition: Terminal Node found!\n";
     }
+    auto new_fusion_id = c10::optional<size_t>(cache_entry.value()->fusion_id);
```
Comment (naoyam) [inline third_party/nvfuser/csrc/python_frontend/fusion_definition.cpp:111]: I get a compilation error as `new_fusion_id` isn't used. Looks like the line below is almost exactly the same except it stores to `fusion_id_`. Can this line be just removed?

Comment (rdspring1) [reply third_party/nvfuser/csrc/python_frontend/fusion_definition.cpp:111]: Yes. It should be removed.

Comment (naoyam) [reply third_party/nvfuser/csrc/python_frontend/fusion_definition.cpp:111]: https://github.com/csarofeen/pytorch/pull/2519

**Pull request merged**